gtkui/mandy.1
.TH mandy 1
.SH NAME
mandy - Mandelbrot set drawer
.SH SYNOPSIS
.B mandy
.RI [ OPTIONS ]
.br
.B mandy
.RI [ OPTIONS ]
.B --draw
.I WIDTH
.I HEIGHT
.I X
.I Y
.I RADIUS
.I MAXITERS
.I PATH
.SH DESCRIPTION
.B mandy
draws images of the Mandelbrot set, with its complement colored
according to the escape time algorithm.
Optionally it displays the corresponding Julia set image.
.SH MOUSE
Double click at a point to zoom in on that point.  Hold control while
double clicking to zoom out.
.PP
Drag to move the displayed image without zooming.
.PP
If the Julia set window is displayed, click or drag with shift held
to, in the Mandelbrot set window, to set the Julia set parameter.
.SH KEYBOARD
Use control with + or minus to zoom in and out respectively.
.SH MENU
Using the
.B File
menu you can save the current Mandelbrot or Julia set image, close the
window, or quit the program.
.PP
The
.B Windows
menu allows you to create and destroy the Julia set window.
.SH "CONTROL PANEL"
The left side of the control panel contains two text entry boxes for
the centre of the
image and a third for the size.
All of these may be edited; press RETURN to commit to the new values.
The size is the radius of the biggest circle that fits on the display,
or (looked at another way) the size of the smaller dimension.
.PP
A fourth box on the left side allows the maximum iteration count to be adjusted.
.PP
The right side shows the position of the pointer and the number of
iterations reached at that point.
.PP
The \fBPrecision\fR dropdown allows the precision to be chosen from
the following:
.TP
.B double
Native floating point type.
.TP
.B simd2
Native floating point type, 2-way SIMD.
Availability depends on platform.
.TP
.B simd4
Native floating point type, 4-way SIMD.
Availability depends on platform.
.TP
.B "long double"
Native floating point type.
Provides more precision than \fBdouble\fR at a small performance cost.
.TP
.B fixed64
64-bit fixed point type.
Provides slightly more precision than \fBdouble\fR at a small
performance cost.
.TP
.B fixed128
128-bit fixed point type.
Provides much more precision than the other types at a relatively
large performance cost.
.PP
Relative performance depends on your hardware and whether the
assembler implementations of the underlying algorithm are used.
.SH OPTIONS
.TP
.B --help\fR, \fB-h
Display a usage message.
.TP
.B --threads \fITHREADS\fR, \fB-t \fITHREADS
Set the number of threads to use.
The default is the number of online CPU cores.
.TP
.B --draw
Draw a Mandelbrot set image and save it to a file.
.SH "OFFLINE DRAWING"
.SS Stills
The
.B --draw
option may be used to draw a single frame and save it to a file.
The syntax is:
.PP
.B mandy
.RI [ OPTIONS ]
.B --draw
.B --
.I WIDTH
.I HEIGHT
.I X
.I Y
.I RADIUS
.I MAXITERS
.I PATH
.TP
.B \(bu
.I WIDTH
and
.I HEIGHT
are the pixel size of the image.
.TP
.B \(bu
.I X
and
.I Y
are the real and imaginary coordinates of the centre of the image and
.I RADIUS
is the radius of the largest inscribed circle (equivalently, the
distance from the centre to the nearest edge).
.TP
.B \(bu
.I MAXITERS
is the maximum iteration count.
.TP
.B \(bu
.I PATH
is the name of the file to save to.
The file type is inferred from the extension, which can be one of
.BR .png ,
.B .ppm
or
.BR .jpeg .
.SS Movies
The
.B --dive
option may be used to create a video of a dive through the set.
The syntax is:
.PP
.B mandy
.RI [ OPTIONS ]
.B --dive
.B --
.I WIDTH
.I HEIGHT
.I START-X
.I START-Y
.I START-RADIUS
.I END-X
.I END-Y
.I END-RADIUS
.I MAXITERS
.I SECONDS
.I PATH
.PP
The radius progresses from \fISTART-RADIUS\fR to \fIEND-RADIUS\fR
exponentially, in other words giving a uniform zoom per frame
transition.
.PP
The centre location progresses from \fISTART-X\fR/\fIY\fR to
\fIEND-X\fR/\fIY\fR linearly.
This may be changed.
Until then, it is recommended that the start and end locations be the
same.
.PP
In addition the following environment variables may be set:
.TP
.B FFMPEG
The encoder to use.
The default is \fBffmpeg\fR.
.TP
.B FRAME_RATE
The frame rate.
The default is \fB25\fR.
.TP
.B BITRATE
The output bitrate.
The default is \fI2097152\fR.
.PP
Note that this option uses a temporary file for each frame in the
current directory.
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>


debian/changelog
mandy (0.0.WIP) unstable; urgency=low

  * Work in progress.

 -- Richard Kettlewell <rjk@greenend.org.uk>  Sun, 07 Nov 2010 11:48:50 +0000



web/mandy.js
window.onload = function() {
    x=-0.75;
    y=0;
    s=2.5;
    f=Math.sqrt(2);
    pstep=64;
    debug=1;
    prune_timeout = 5 * 60;
    prune_interval = 60;
    queue_latency_ms = 10;

    dragging=false;
    generation=0;
    mcanvas = document.getElementById('m');
    ucanvas = document.getElementById('u');
    cache={};
    queue=[];
    mctx = mcanvas.getContext('2d');
    uctx = ucanvas.getContext('2d');

    resize();
    compute();
    render(x, y);
    run_queue(16);

    ucanvas.addEventListener("mousedown", mousedown);
    ucanvas.addEventListener("mouseup", mouseup);
    ucanvas.addEventListener("mousemove", mousemove);
    ucanvas.addEventListener("mouseleave", mouseleave);
    ucanvas.addEventListener("dblclick", dblclick);
    ucanvas.addEventListener("wheel", wheel);
    window.onresize = window_resized;
    window.setInterval(prune, prune_interval * 1000)
}

// Adjust the canvas to match the containing window
function resize() {
    ucanvas.width = mcanvas.width = window.innerWidth;
    ucanvas.height = mcanvas.height = window.innerHeight;
}

// Compute useful things
function compute() {
    // Find the canvas size
    w = mcanvas.width;
    h = mcanvas.height;
    // Find the size in complex plane units
    if(w >= h) {
        // Wider than high (or equal).  Treat s as the height.
        cw = s * w / h;
        ch = s;
    } else {
        // Higher than wide.  Treat s as the width.
        cw = s;
        ch = s * h / w;
    }
    // Find the bottom left corner in complex plane
    cxc = x - cw / 2;
    cyc = y - ch / 2;
    // Find the step in complex plane units
    cstep = pstep * cw / w;
    // Attempt to align to a consistent grid
    cxo = cxc % cstep;
    if(cxo < 0)
        cxo += cstep;
    cxb = cxc - cxo;
    pxs = -Math.floor(cxo * pstep / cstep);
    cyo = cyc % cstep;
    if(cyo < 0)
        cyo += cstep;
    cyb = cyc - cyo;
    pys = -Math.floor(cyo * pstep / cstep);
    if(debug) {
        console.log("pxs:", pxs, "pys:", pys, "w:", w, "h:", h,
                    "pstep:", pstep,
                    "cxb:", cxb, "cyb:", cyb, "cstep: ", cstep);
    }
}

// Render the whole canvas
function render(ccx, ccy) {
    var px, py, pyr, img;
    var cx, cy;
    var n;
    var i, jobs, job;
    n = 0;
    // Start a new generation
    ++generation;
    if(debug > 1) {
        mctx.fillStyle='#00ffff';
        mctx.fillRect(0, 0, w, h);
        mctx.font="12px serif";
    }
    // Visit every square in the grid in any old order
    jobs = []
    for(px = pxs, cx = cxb; px < w; px += pstep, cx += cstep) {
        for(pyr = pys, cy = cyb; pyr < h; pyr += pstep, cy += cstep) {
            py = h - pyr - pstep;
            if(debug > 1) {
                console.log("px: ", px, " py: ", py, " cx: ", cx, " cy: ", cy);
                mctx.fillStyle='rgb('+n+','+n+','+n+')';
                mctx.fillRect(px, py, pstep, pstep);
                mctx.fillStyle='#ff0000'
                mctx.fillText(cx + "," + cy, px + 2, py + 18);
                n = (n + 4) % 256;
            }
            job = [x,y,cx,cy,px,py,ccx,ccy,cstep,pstep,generation]
            jobs.push(job);
        }
    }
    jobs.sort(comparator)
    queue = queue.concat(jobs)
    annotate();
}

function annotate() {
    uctx.font="12px serif";
    uctx.fillStyle = '#ffffff';
    uctx.fillRect(w - 128 - 0.5,
                  h - 64 - 0.5,
                  128, 64);
    uctx.fillStyle = '#000000';
    uctx.textAlign = 'end';
    uctx.fillText("x:", w - 128 + 4 + 16, h - 64 + 4 + 16);
    uctx.fillText("y:", w - 128 + 4 + 16, h - 64 + 4 + 32);
    uctx.fillText("s:", w - 128 + 4 + 16, h - 64 + 4 + 48);
    uctx.textAlign = 'start';
    uctx.fillText(x, w - 128 + 4 + 20, h - 64 + 4 + 16);
    uctx.fillText(y, w - 128 + 4 + 20, h - 64 + 4 + 32);
    uctx.fillText(s, w - 128 + 4 + 20, h - 64 + 4 + 48);
    uctx.strokeStyle = '#000000';
    uctx.strokeRect(w - 128 - 0.5, h - 64 - 0.5, 128, 64);
    uctx.strokeStyle = '#ffffff';
    uctx.strokeRect(w - 127 - 0.5, h - 63 - 0.5, 126, 62);
}

// Run the queue
function run_queue(active_max) {
    var job;
    if(queue.length > 0) {
        active = 0;
        early_skip = 0;
        if(debug)
            console.log("queue run", queue.length);
        while(queue.length > 0 && active < active_max) {
            job = queue.shift();
            run(job);
        }
        if(debug && early_skip > 0)
            console.log("early skip", early_skip);
    }
}

// Run the queue in a little while
function run_queue_soon() {
    if(queue.length > 0) {
        if(debug)
            console.log("deferring queue run", queue.length);
        window.setTimeout(function() {
            run_queue(16);
        }, queue_latency_ms);
    }
}

function run(job) {
    var gen = job[10];
    if(gen == generation) {
        var cx = job[2], cy=job[3];
        var px = job[4], py = job[5];
        var cstep = job[8], pstep = job[9];
        var url="/cgi-bin/mandy.fcgi?x="+cx+"&y="+cy+"&s="+cstep+"&w="+pstep+"&h="+pstep;
        var img = url in cache ? cache[url] : new Image();
        img.used = new Date().valueOf();
        if(img.src !== "" && img.complete)
            mctx.drawImage(img, px, py);
        else {
            cache[url] = img;
            img.addEventListener("load", function() {
                if(gen === generation) {
                    mctx.drawImage(img, px, py);
                    run_queue(1);
                } else if(debug > 1)
                    console.log("late skip at", px, py);
            }, false);
            if(img.src === "")
                img.src = url;
            ++active;
        }
    } else
        ++early_skip;
}

function distance(job) {
    var x, y, cx, cy, dx, dy;
    cx = job[2];
    cy = job[3];
    ccx = job[6];
    ccy = job[7];
    dx = (ccx - cx);
    dy = (ccy - cy);
    return dx * dx + dy * dy;
}

function comparator(a, b) {
    var da = distance(a);
    var db = distance(b);
    if(da < db)
        return -1;
    else if(da > db)
        return 1;
    else
        return 0;
}

// Scale up or down
function scale(pcx, pcy, ccx, ccy, k) {
    //    cxc = x - cw / 2
    //    ccx = cxc + pcx * cstep / pstep
    //        = cxc + pcx * (pstep * cw / w) / pstep
    //        = cxc + pcx * cw / w
    //        = x - cw/2 + pcx * cw/w
    // =>   x = ccx + cw/2 - pcx * cw/w
    // We want to fix ccx and pcx while replacing x and cw, so:
    //     x' = ccx + cw'/2 - pcx * cw'/w
    //        = ccx + k*cw/2 - k*pcx*cw/w
    //        = ccx + k*cw*(0.5-pcx/w)
    x = ccx + k * cw * (0.5-pcx/w);
    y = ccy + k * ch * (0.5-pcy/h);
    s *= k;
    compute();
    render(ccx, ccy);
    run_queue_soon();
}

// Drag (translate)
function drag(from, to) {
    var xd = to[0] - from[0];
    var yd = to[1] - from[1];
    if(xd == 0 && yd == 0)
        return;
    if(debug)
        console.log("drag by ", xd, " x ", yd);
    x -= xd * cw / w;
    y -= yd * ch / h;
    compute();
    render(x, y);
    run_queue_soon();
}

// Find where a click event landed
function where(e) {
    var rect = e.target.getBoundingClientRect();
    return [e.clientX - rect.left,
            h + rect.top - e.clientY];
}

// Prune cache entries that haven't been used in a while
function prune() {
    var k, i, prunable = [], limit;
    limit = new Date().valueOf() - prune_timeout * 1000;
    for(k in cache)
        if(cache[k].used < limit)
            prunable.push(k);
    if(prunable.length > 0) {
        if(debug)
            console.log("pruning", prunable.length, "items from cache");
        for(i in prunable)
            delete cache[prunable[i]];
        if(debug)
            console.log("cache size now", Object.keys(cache).length);
    }
}

function mousedown(e) {
    draglast = where(e);
    dragging = true;
}

function mousemove(e) {
    if(dragging) {
        dragnew = where(e);
        drag(draglast, dragnew);
        draglast = dragnew;
    }
}

function mouseup(e) {
    dragging = false;
}

function mouseleave(e) {
    dragging = false;
}

function dblclick(e) {
    var clicked = where(e);
    var pcx = clicked[0], pcy = clicked[1];
    var ccx = cxc + pcx * cstep / pstep;
    var ccy = cyc + pcy * cstep / pstep;
    dragging = false;           // just in case
    if(debug)
        console.log("clicked", clicked, [ccx, ccy]);
    scale(pcx, pcy, ccx, ccy, 1/f);
}

function wheel(e) {
    var clicked = where(e);
    dragging = false;
    if(e.deltaY == 0)
        return;
    var pcx = clicked[0], pcy = clicked[1];
    var ccx = cxc + pcx * cstep / pstep;
    var ccy = cyc + pcy * cstep / pstep;
    var k;
    dragging = false;           // just in case
    switch(e.deltaMode) {
    case 0:                     // pixel
        if(Math.abs(e.deltaY) < 20)
            k = Math.pow(f, 0.25);
        else
            k = f;
        break;
    default:
        k = f;
        break;
    }
    if(e.deltaY > 0)
        k = 1/k;
    if(debug)
        console.log("wheel", clicked, [ccx, ccy], k);
    scale(pcx, pcy, ccx, ccy, k);
}

function window_resized(e) {
    console.log("window_resized");
    resize();
    compute();
    render(x, y);
    run_queue_soon();
}
